lddwrap/__init__.py

```diff
@@ -0,0 +1,228 @@
+            self.soname, self.path, self.found, self.mem_address, self.unused)
+
+
+def parse_line(line: str) -> Dependency:
+    """
+    Parse single line of ldd output.
+
+    :param line: to parse
+    :return: dependency or none if line was empty
+
+    """
+    found = not 'not found' in line
+    parts = [part.strip() for part in line.split(' ')]
+
+    # There are two types of outputs for a dependency, with or without soname. For example:
+    # with soname: 'libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f9a19d8a000)'
+    # without soname: '/lib64/ld-linux-x86-64.so.2 (0x00007f9a1a329000)'
+    # with soname but not found: 'libboost_program_options.so.1.62.0 => not found'
+    # with soname but without rpath: 'linux-vdso.so.1 =>  (0x00007ffd7c7fd000)'
+    if '=>' in line:
+        if len(parts) != 4:
+            raise RuntimeError("Expected 4 parts in the line but found {}: {}".format(len(parts), line))
+
+        soname = None
+        dep_path = None
+        mem_address = None
+        if found:
+            soname = parts[0]
+            if parts[2] != '':
+                dep_path = pathlib.Path(parts[2])
+            mem_address = parts[3]
+        else:
+            if "/" in parts[0]:
+                dep_path = pathlib.Path(parts[0])
+            else:
+                soname = parts[0]
+
+        return Dependency(soname=soname, path=dep_path, found=found, mem_address=mem_address)
+    else:
+        if len(parts) != 2:
+            raise RuntimeError("Expected 2 parts in the line but found {}: {}".format(len(parts), line))
+
+        return Dependency(soname=None, path=pathlib.Path(parts[0]), found=True, mem_address=parts[1])
+
+
+def get_dependencies(path: pathlib.Path,
+                     verbose: bool = False,
+                     unused: bool = False,
+                     env: Optional[Dict[str, str]] = None) -> List[Dependency]:
+    """
+    Retrieve a list of dependencies of the given binary.
+
+    :param path: path to a file
+    :param verbose: additionally print all information
+    :param unused: additional adds information if dependency is used
+    :param env: the environment to use
+    :return: list of dependencies
+
+    Options copied from ldd documentary:
+    --version
+              Print the version number of ldd.
+
+       -v, --verbose
+              Print all information, including, for example, symbol
+              versioning information.
+
+       -u, --unused
+              Print unused direct dependencies.  (Since glibc 2.3.4.)
+
+       -d, --data-relocs (not implemented)
+              Perform relocations and report any missing objects (ELF only).
+
+       -r, --function-relocs (not implemented)
+              Perform relocations for both data objects and functions, and
+              report any missing objects or functions (ELF only).
+
+       --help Usage information. (not implemented)
+    """
```

Comment:
Comment C:

This part of the code seems to be a bit overcomplicated. How about writing it as follows:

```
if not path.is_file():
        raise IOError("Not a file: {}".format(path))

commands = ["ldd"]  # type: List[str]
if verbose:
  commands.append("--verbose")

commands.append(path.as_posix())

proc = subprocess.Popen(commands, stdout=subprocess.PIPE, stderr=subprocess.PIPE, universal_newlines=True, env=env)

out, err = proc.communicate()
if proc.returncode != 0:
    raise RuntimeError("Failed ldd-ing external libraries of {} with code {}:\nout:\n{}\n\nerr:\n{}".format(
        path, proc.returncode, out, err))

dependencies = dependencies_parse_helper() # function which you would have to write, terminates at the end of the file or at "Version information"

if unused:
   ...
   dependencies = update_unused(dependencies, out_unused)

return dependencies
```

This way you would avoid the confusing recursive call of `get_dependencies` function

Reply:
Thanks! Your idea is way easier.